### DISCUSSION TITLE
New SSP model from gluing BC03 and XSL

### DISCUSSION THREAD

**Discussion by Pedro-Floriano** [mboquien/cigale#87] [Ideas]
Hello everyone, for the last months I've been looking for SSP models that would help us improving our study on SED fitting using J-PLUS photometry of Globular Clusters. After considerations we decided that the best option was to create a new SSP model from gluing the models BC03 (Padova2000) and XSL (P00 Kroupa) together. So here are the details:

This model have a range of 27 ages: [8.91e+08, 1.00e+09, 1.12e+09, 1.26e+09, 1.41e+09, 1.58e+09, 1.78e+09, 2.00e+09, 2.24e+09, 2.51e+09, 2.82e+09, 3.16e+09, 3.55e+09, 3.98e+09, 4.47e+09, 5.01e+09, 5.62e+09, 6.31e+09, 7.08e+09, 7.94e+09, 8.91e+09, 1.00e+10, 1.12e+10, 1.26e+10, 1.41e+10, 1.59e+10, 1.78e+10]

And a range of metallicities: Z = [0.0004, 0.001, 0.004, 0.008, 0.019, 0.03] 

The collage was done by gluing the more defined parts of each model. So in this case, the first part comes from BC03, the middle part from XSL, and the last part also from BC03. I'll append some examples of the spectra including a comparison of the BC03 model alone, versus the glued model. 
I'd love if this model could be added to the code in some form so we could run some tests, I would include here the file with model but it is 99MB (after compressed into a zip file) so it wasn't supported, the best I can do for now is to share the link for the drive: [https://drive.google.com/file/d/1SUZu0esca_qpWn37jDcjY4796Ikvpkrh/view?usp=sharing](url)

After that, you can open the file with the following commands:
```
with open('path_to_file', 'rb') as handle:
    glued_spectra = pickle.load(handle)
```

![image](https://github.com/mboquien/cigale/assets/113158126/fd211cc2-0dd7-4e8e-8edf-ab0bd166f7de)
![image](https://github.com/mboquien/cigale/assets/113158126/9fc6cbc8-d09f-47f4-babc-7ee3cc1508e1)

![image](https://github.com/mboquien/cigale/assets/113158126/12dd1822-e5f1-46bd-9931-23bb6110da42)
![image](https://github.com/mboquien/cigale/assets/113158126/8594c1fb-b473-44e1-9336-bc1a85a105a9)

Thank you, in advance.

**Comment by Pedro-Floriano** [comment]
One thing I forgot to mention, the model goes from 91nm to 1.4e+5nm. The plots are just up to 3000nm because after that the spectra is just from BC03, which appears to be a flat line until the end.

**Comment by mboquien** [comment]
Could you share the models in another format, for instance a FITS file? I fear pickle files are not designed for exchanging data. Just a comment from a quick glance at the plots, it seems there is a lot of noise around 1.4 and 1.9 μm, which very much look like sky line remnants. I would suggest to replace these sections of the spectra as it could generate some issues.

**Reply by Pedro-Floriano** [reply]
Thank you answering, I'll have a look on the noise you mentioned. For now I can share here the models in FITS format: [https://drive.google.com/drive/folders/1STExBr50QCcT5XJ8idoi2Y0uyxE-Xlgq?usp=sharing](url)

**Reply by Pedro-Floriano** [reply]
About the noise in the infrared, I was talking with my collaborators and we agreed that as we are using JPLUS photometry, it won't affect the results because the filters only go up to 1μm. In the future, thinking about other's research we might improve the infrared part of these models, but for now we would much appreciate if we could run some tests using these models in cigale.

**Reply by Pedro-Floriano** [reply]
Hi @mboquien, have you got some time to have a look at this implementation?, and if not, is there a way for us to implement it ourselves just to run some tests?. We are looking into some new data from objects in the JPLUS field, and we would like to run it in cigale with these models.

**Reply by mboquien** [reply]
I fear I am starting a new class this month, leaving me with little free time. What I would suggest is to use one of the existing models as a template for the implementation. If you would like, we can organize a telecon to get you going.

**Reply by Pedro-Floriano** [reply]
A telecon would be perfect. I read the template for the BC03 models, and they seem a little complicated to reach when starting from the way we organized our model. But if we could arrange a meeting we can talk more about it. Any time would do for us. Thank you.

**Comment by Pedro-Floriano** [comment]
Hey @mboquien, so I went in the cigale database and found the Maraston2005 models, and saw how they were written, then I tried to write the model we have in that format. Here it is: [https://drive.google.com/drive/folders/1yKdwgB_2lf0U2j4v_J0_hud0lF7gq7mq?usp=sharing](url)
My question now is how can I replace the m2005 files for these ones to run cigale with?
Thank you.

**Reply by mboquien** [reply]
The idea would not to replace the Maraston models but to add yours. You can duplicate the m2005 module (and related files for the import) with a new name (for instance `xsl`). Then if should just be some light adaptations.

**Reply by Pedro-Floriano** [reply]
Ok, so I have the `cigale-v2022.1` file, inside there is the `database_builder` (where I added the xsl folder) with the `__init__.py`, I edited this py file by duplicating the m2005 part of the code and changing the parts that explicit say m2005 to xsl. Now, do I need to rebuild the code, and if so, how do I do that? Or I can just start running it?

**Reply by mboquien** [reply]
To build the database you need to run `python setup.py build`. Just make sure it is called in [build_base()](https://gitlab.lam.fr/cigale/cigale/-/blob/v2022.1/database_builder/__init__.py?ref_type=tags#L882). After the database/code are installed it should then work directly.

**Comment by Pedro-Floriano** [comment]
Hi @mboquien, I've been trying to do as you said, but I'm getting an error that I don't really understand when running `python setup.py build`, here is the full error message:
[error.txt](https://github.com/user-attachments/files/15683290/error.txt)

`ValueError: cannot reshape array of size 0 into shape (0,newaxis)`


I don't really understand why this is happening because my data is written exactly the same way as maraston's, so I just copied the way maraston is built and changed the metallicites to fit my data. The only difference between my data and maraston's is the number of lines in the files, mine has around 1320000 lines while maraston has around 80000 lines.

**Reply by Pedro-Floriano** [reply]
By setting checkpoints in the `__init__.py` I found that the error occurs when this line is read:
`mass_table = mass_table[1:7, mass_table[0] == metallicity]` for some reason this is turning the mass_table into an empty list.
So when this line is read:
`mass_table = interpolate.interp1d(mass_table[0] * 1e3, mass_table[1:], assume_sorted=True)(fine_time_grid)`
It gives the error.
This is the '__init__py' I'm using: 
[init.txt](https://github.com/user-attachments/files/15686290/init.txt)

**Reply by mboquien** [reply]
It is probably that the time grid is not the same. You need to adjust `fine_time_grid` (towards the beginning of your function) to your case.

**Reply by Pedro-Floriano** [reply]
I'll try to fix that, but even if this is wrong, isn't it weird that the mass_table gets empty right after running this: `mass_table = mass_table[1:7, mass_table[0] == metallicity]` because before that, it had values in it

**Reply by Pedro-Floriano** [reply]
Hi @mboquien I'm still having some troubles, maybe can we arrange a meeting this next week to have a look over this code?

**Reply by mboquien** [reply]
Sure. Write me an email, we can arrange this.

**Reply by Pedro-Floriano** [reply]
Hi @mboquien, in our last meeting we talked about how the mass_table was affecting the build of the code. As the metallicities in XSL are different from the metallicities in M2005. Also we talked about how our studies don't require any type of mass estimation, so I took out everything related to this mass_table. Still when I run the build, I just get a "Killed" message and the run is stopped. I went through the files in cigale and inside the folder named "data" there are folders named after all the models (bc03, 2005, dl2007...) and after trying to run the build it seems that it created a folder for xsl, but inside there is only a file named "parameters.pickle", inside the other folders, I found this "parameters.pickle" but also other pickle files related to their imf.

So my questions are: is there a problem to try and build this xsl model without the mass_table file? also, in this folder named "data", how are the pickle files generated?, and how can I generate this files for xsl?

**Reply by Pedro-Floriano** [reply]
In the next weeks I'll get on looking into it more closely, but I wanted to leave here where I'm standing.

**Reply by mboquien** [reply]
It is difficult to say much without seeing the log. That the process is killed suggest you are exhausting the RAM though.

**Reply by Pedro-Floriano** [reply]
Hello @mboquien, I was out for a conference. I recently returned, made some changes in the module, mainly reducing its size so the RAM isn't exhausted anymore. And it worked, I ran `python setup.py build` and it finished without any problem, then I ran `python setup.py develop --user` which also went ok. But now I'm having some trouble using the module in runs with cigale. When trying to run `pcigale genconf` with xsl, I get the following error:
![image](https://github.com/user-attachments/assets/1b769b54-8166-4dc8-b39b-701dcd8326d1)
![image](https://github.com/user-attachments/assets/ee379948-38b8-4a56-87c4-6ec70d27c64d)

**Reply by Pedro-Floriano** [reply]
Even though, I have the xsl.py created in the pcigale/sed_modules
![image](https://github.com/user-attachments/assets/fff3e09f-3f33-4363-81a6-bf944321b23a)
The xsl directory that was created in pcigale/data using `python setup.py build` also seems ok
![image](https://github.com/user-attachments/assets/9d124192-7fd4-4666-96cc-6fe042b5a992)

**Reply by Pedro-Floriano** [reply]
Was there anything else I needed to do before trying to run the code?

**Reply by mboquien** [reply]
I think it does not find the module because you are running a local regular installation. You can tell from the path of the files, they are in `~/.local/`. Also not that it is using python 3.8, which is a little old. You may want to look into this so that it actually runs the development installation.

**Reply by Pedro-Floriano** [reply]
So I'll look into updating python. But meanwhile could you give me some ideas on how to install the code differently, so its not a local installation?

**Reply by mboquien** [reply]
A local installation is fine. I think you are just mixing different versions of python and cigale. I would recommend you consult with your local sysadmin to set that straight.
